fms/models/__init__.py

```diff
@@ -65,6 +68,48 @@ def list_variants(architecture: str):
     return list(__models[architecture].keys())
 
 
+def __maybe_infer_model_variant(
+    architecture: str,
+    variant: str,
+    model_path: Optional[str],
+    source: Optional[str],
+    **kwargs,
+) -> Tuple[str, str, Optional[str], Optional[str], Dict[str, Any]]:
+    """Infer the model variant configuration from different sources, currently only supported sources are hf"""
+    extra_kwargs = kwargs
+
+    if architecture in ("hf_pretrained", "hf_configured"):
+        logger.info(f"inferring model configuration from {variant}")
+
+    if architecture == "hf_pretrained":
+        if len(kwargs) > 0:
+            logger.warning(
+                f"ignoring the following parameters as a pretrained model with an inferred configuration is being loaded: {list(kwargs.keys())}"
+            )
+
+        if model_path is not None or source is not None:
+            raise ValueError(
+                """architecture="hf_pretrained" implies model weights will be downloaded and extracted from hf cache and loaded into the model, therefore model_path and source should not be set"""
+            )
+
+        from fms.models.hf.utils import _infer_model_configuration  # type: ignore
+
+        extra_kwargs = _infer_model_configuration(variant, download_weights=True)
+        architecture = extra_kwargs.pop("architecture")
+        variant = extra_kwargs.pop("variant")
+        model_path = extra_kwargs.pop("model_path")
+        source = "hf"
+    elif architecture == "hf_configured":
+        from fms.models.hf.utils import _infer_model_configuration  # type: ignore
+
+        extra_kwargs = _infer_model_configuration(variant, download_weights=False)
+        extra_kwargs = {**extra_kwargs, **kwargs}
+        architecture = extra_kwargs.pop("architecture")
+        variant = extra_kwargs.pop("variant")
```

Comment:
a lot of copy/paste here, I think most of these can just be done once for both cases? `if architecture in('hf_pretrained', 'hf_config'): extra_kwargs = _infer_model_config(variant, download_weights=architecture=='hf_pretrained')` ...etc

Reply:
Fixed in latest commit